packages/flutter/lib/src/services/text_input.dart

```diff
@@ -671,6 +686,370 @@ class RawFloatingCursorPoint {
+///
+/// This class should not be used directly, and should be extended for different
+/// types of deltas.
+abstract class TextEditingDelta with TextEditingDeltaUtils {
+  /// Creates a delta for a given change to the editing state.
+  ///
+  /// {@template flutter.services.TextEditingDelta}
+  /// The [oldText], [deltaText], [deltaRange], [selection], and [composing]
+  /// arguments must not be null.
+  /// {@endtemplate}
+  const TextEditingDelta({
+    required this.oldText,
+    required this.deltaText,
+    required this.deltaRange,
+    required this.selection,
+    required this.composing,
+  }) : assert(oldText != null),
+       assert(deltaText != null),
+       assert(deltaRange != null),
+       assert(selection != null),
+       assert(composing != null);
+
+  /// Creates an instance of this class from a JSON object by checking the
+  /// deltaType sent by the engine and building the appropriate delta.
+  factory TextEditingDelta.fromJSON(Map<String, dynamic> encoded) {
+    switch (encoded['deltaType'] as String) {
+      case 'TextEditingDeltaType.insertion':
+        return TextEditingDeltaInsertion.fromJSON(encoded);
+      case 'TextEditingDeltaType.deletion':
+        return TextEditingDeltaDeletion.fromJSON(encoded);
+      case 'TextEditingDeltaType.replacement':
+        return TextEditingDeltaReplacement.fromJSON(encoded);
+      case 'TextEditingDeltaType.equality':
+        return TextEditingDeltaEquality.fromJSON(encoded);
+      default:
+        return TextEditingDeltaEquality.fromJSON(encoded);
+    }
+  }
+
+  /// The old text state before the delta has occured.
+  final String oldText;
+
+  /// The value represents the text that is being inserted/deleted by this delta.
+  ///
+  /// This value will slightly vary based on the [TextEditingDeltaType].
+  ///
+  /// For a [TextEditingDeltaType.insertion] this will be the character/s being
+  /// inserted.
+  ///
+  /// For a [TextEditingDeltaType.deletion] this will be the character/s being
+  /// deleted.
+  ///
+  /// For a [TextEditingDeltaType.replacement] this will be the text that is
+  /// replacing the [TextEditingDelta.deltaRange].
+  ///
+  /// For a [TextEditingDeltaType.equality] this will be an empty string.
+  final String deltaText;
+
+  /// This value can either represent a range of text that the delta is changing
+  /// or if is a collapsed range then it represents the point where this delta
+  /// began.
+  ///
+  /// This value will slightly vary based on the [TextEditingDeltaType].
+  ///
+  /// For a [TextEditingDeltaType.insertion] this will be a collapsed range
+  /// representing the cursor position where the insertion began.
+  ///
+  /// For a [TextEditingDeltaType.deletion] this will be the range of text
+  /// that was deleted.
+  ///
+  /// For a [TextEditingDeltaType.replacement] this will be the range of
+  /// characters that are being replaced.
+  ///
+  /// For a [TextEditingDeltaType.equality] this will be a collapsed range
```

Comment:
If a property does not make sense in a subtype then it may make sense to remove it from the common interface. Same for `deltaText`.